File: 2008/J081.pas
var
  vars: array['A'..'Z'] of longint;
  i, n, first, second, pos: longint;
  s: string;
  j: char;
  add: boolean;
begin
  for j := 'A' to 'Z' do
    vars[j] := -1000001;
  readln(n);
  for i := 1 to n do
  begin
    readln(s);
    s[length(s) + 1] := ' ';
    if (s[3] >= 'A') then
    begin
      first := vars[s[3]];
      pos := 4
    end
    else
    begin
      first := 0;
      pos := 3;
      if (s[3] = '-') then
        pos := 4;
      while ((s[pos] >= '0') and (s[pos] <= '9')) do
      begin
        first := first * 10 + ord(s[pos]) - ord('0');
        inc(pos)
      end;
      if (s[3] = '-') then
        first := -first
    end;
    if (s[pos] = ' ') then
    begin
      vars[s[1]] := first;
      continue
    end;
    add := s[pos] = '+';
    inc(pos);
    if (s[pos] >= 'A') then
      second := vars[s[pos]]
    else
    begin
      second := 0;
      if (s[pos] = '-') then
      begin
        add := not add;
        inc(pos)
      end;
      while ((s[pos] >= '0') and (s[pos] <= '9')) do
      begin
        second := second * 10 + ord(s[pos]) - ord('0');
        inc(pos)
      end;
    end;
    if (add) then
      vars[s[1]] := first + second
    else
      vars[s[1]] := first - second
  end;
  for j := 'A' to 'Z' do
    if (vars[j] <> -1000001) then
      writeln(j, '=', vars[j])
end.

File: 2008/J081.cpp
#include <cstdio>
char s[50];
int vars[26];
int main() {
  for (int i = 0; i < 26; i++)
    vars[i] = -1000001;
  int n;
  scanf("%d", &n);
  for (int i = 0; i < n; i++) {
    int first, second, pos;
    bool add;
    scanf("%s", s);
    if (s[2] >= 'A') {
      first = vars[s[2] - 'A'];
      pos = 3;
    } else {
      first = 0;
      pos = 2;
      if (s[2] == '-')
        pos++;
      while (s[pos] >= '0' && s[pos] <= '9')
        first = first * 10 + (s[pos++] - '0');
      if (s[2] == '-')
        first *= -1;
    }
    if (s[pos] == 0) {
      vars[s[0] - 'A'] = first;
      continue;
    }
    add = s[pos++] == '+';
    if (s[pos] >= 'A') {
      second = vars[s[pos] - 'A'];
    } else {
      second = 0;
      if (s[pos] == '-') {
        add = !add;
        pos++;
      }
      while (s[pos] >= '0' && s[pos] <= '9')
        second = second * 10 + (s[pos++] - '0');
    }
    vars[s[0] - 'A'] = first + (add ? second : -second);
  }
  for (int i = 0; i < 26; i++)
    if (vars[i] != -1000001)
      printf("%c=%d\n", 'A' + i, vars[i]);
  return 0;
}
